src/main/java/com/snowflake/kafka/connector/internal/ingestsdk/StreamingClientManager.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2023 Snowflake Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.snowflake.kafka.connector.internal.ingestsdk;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.snowflake.kafka.connector.Utils;
+import com.snowflake.kafka.connector.internal.LoggerHandler;
+import com.snowflake.kafka.connector.internal.SnowflakeErrors;
+import com.snowflake.kafka.connector.internal.streaming.StreamingUtils;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Provides access to the streaming ingest clients. This should be the only place to manage clients.
+ */
+public class StreamingClientManager {
+  private LoggerHandler LOGGER;
+
+  private Map<Integer, KcStreamingIngestClient> taskToClientMap;
+  private int maxTasks;
+  private final int minTasks = 0;
+
+  // TESTING ONLY - inject the client map
+  @VisibleForTesting
+  public StreamingClientManager(Map<Integer, KcStreamingIngestClient> taskToClientMap) {
+    this();
+    this.taskToClientMap = taskToClientMap;
+  }
+
+  /** Creates a new client manager */
+  protected StreamingClientManager() {
+    LOGGER = new LoggerHandler(this.getClass().getName());
+    this.taskToClientMap = new HashMap<>();
+    this.maxTasks = 0;
+  }
+
+  /**
+   * Creates as many clients as needed with the connector config and kc instance id. This assumes
+   * that all taskIds are consecutive ranging from 0 -> maxTasks.
```

Comment:
Is this guaranteed to hold true? What happens if they aren't consecutive?

Reply:
In my experience they are consecutive, but this is a very good point. I thiiiink we set the taskid internally ourselves, but overall I definitely think we need a better way of keeping track the objects and corresponding identifiers for them

Reply:
I'll dig into the code to confirm it holds true

Reply:
We add the task_id property consecutively starting from 0 to the config in SnowflakeSinkConnector.taskConfigs. This property is then passed to the task with the parsedConfig at SnowflakeSinkTask.start(parsedConfig). Not ideal, but we are guaranteed consecutive task ids